File: static/midi.js
// ===================================================================
// MIDI — in, mapped, and onto the picture
// ===================================================================
// Mirrors src/midi.py. Both sides parse the same bytes and match the same
// bindings, so a mapping behaves identically whether the message arrived from a
// port in this browser or over HTTP from another application.
//
// Nothing here assumes hardware. Web MIDI is asked for and may be refused —
// no browser support, no permission, no devices — and every refusal is reported
// rather than swallowed, because "no activity" and "no connection" look
// identical on screen and only one of them is a problem.

const MIDI_RESERVED_CC = [20, 21, 22, 23]; // rad's speed axes; watch, never reuse

function midiParse(bytes) {
    const data = Array.from(bytes || []);
    if (!data.length) throw new Error('no bytes');

    const status = data[0];
    if (status < 0x80) throw new Error(`0x${status.toString(16)} is a data byte`);

    if (status >= 0xF8) {
        const transport = { 0xF8: 'clock', 0xFA: 'start', 0xFB: 'continue', 0xFC: 'stop' };
        if (!transport[status]) throw new Error(`unsupported system message`);
        return { type: transport[status] };
    }
    if (status >= 0xF0) throw new Error('system common is not mapped');

    const kind = status & 0xF0;
    const channel = (status & 0x0F) + 1;
    const d1 = data[1];
    const d2 = data[2];

    switch (kind) {
        case 0x90:
            // Velocity zero is a note-off. Devices send this, and reading it as
            // a note-on leaves the device lit with nothing playing.
            return { type: d2 ? 'note_on' : 'note_off', channel, note: d1, value: d2 };
        case 0x80:
            return { type: 'note_off', channel, note: d1, value: d2 };
        case 0xA0:
            return { type: 'aftertouch', channel, note: d1, value: d2 };
        case 0xB0:
            return { type: 'cc', channel, controller: d1, value: d2 };
        case 0xC0:
            return { type: 'program', channel, value: d1 };
        case 0xD0:
            return { type: 'aftertouch', channel, value: d1 };
        case 0xE0:
            return { type: 'pitchbend', channel, value: (d2 << 7) | d1 };
        default:
            throw new Error(`unhandled status 0x${status.toString(16)}`);
    }
}

function midiIntensity(message) {
    if (message.type === 'note_off' || message.type === 'stop') return 0;
    if (message.type === 'pitchbend') return (message.value || 0) / 16383;
    if (message.value === undefined || message.value === null) return 1;
    return message.value / 127;
}

// A field left out is a wildcard, so a binding survives a machine being moved
// to another channel.
function midiMatches(source, message) {
    if (source.channel != null && message.channel !== source.channel) return false;

    switch (source.type) {
        case 'channel':
            return message.channel != null;
        case 'note':
            if (!['note_on', 'note_off', 'aftertouch'].includes(message.type)) return false;
            return source.note == null || message.note === source.note;
        case 'cc':
            if (message.type !== 'cc') return false;
            return source.controller == null || message.controller === source.controller;
        case 'program':
            return message.type === 'program';
        case 'pitchbend':
            return message.type === 'pitchbend';
        case 'transport':
            return ['clock', 'start', 'stop', 'continue'].includes(message.type);
        default:
            return false;
    }
}

// All matching bindings fire. One message can genuinely concern two devices —
// a clock lights everything following it — and picking one winner would hide
// that.
function midiRoute(bindings, message) {
    return (bindings || [])
        .filter(binding => midiMatches(binding.source, message))
        .map(binding => ({
            binding: binding.id,
            module: binding.module,
            parameter: binding.parameter || null,
            jack: binding.jack || null,
            intensity: midiIntensity(message),
            message,
        }));
}

// ===================================================================
// THE PORT
// ===================================================================
class MidiInput {
    constructor({ onActivity, onStatus, onMessage }) {
        this.onActivity = onActivity;
        this.onStatus = onStatus || (() => {});
        // Every message, matched or not. Learn needs the ones nothing matches —
        // that is the whole point of binding something new.
        this.onMessage = onMessage || (() => {});
        this.access = null;
        this.ports = [];
        this.bindings = [];
        this.lastMessage = null;
        // Counted rather than inferred: a rack with no bindings and a rack with
        // no cable both show nothing, and the count tells them apart.
        this.received = 0;
        this.matched = 0;
    }

    setBindings(bindings) {
        this.bindings = bindings || [];
        return this.bindings;
    }

    get available() {
        return typeof navigator !== 'undefined' && Boolean(navigator.requestMIDIAccess);
    }

    // Why MIDI will not work here, in words that name the actual cause.
    //
    // `available` was answering one question with two meanings. Web MIDI is only
    // exposed in a secure context, so on `http://192.168.1.151:8000` -- the LAN
    // address this server prints at startup, and the one the onboarding page
    // recommends for an on-device test -- `navigator.requestMIDIAccess` is
    // simply absent, and the app said "this browser has no Web MIDI". The
    // browser has it. The page is not a secure context. Measured in Chromium:
    // loopback reports `isSecureContext: true` and the API present; the LAN
    // address reports false and absent.
    diagnose() {
        if (typeof navigator === 'undefined') {
            return { state: 'unsupported', detail: 'no navigator in this runtime' };
        }
        if (typeof window !== 'undefined' && window.isSecureContext === false) {
            const here = (typeof location !== 'undefined' && location.origin) || 'this address';
            return {
                state: 'insecure-context',
                detail: `Web MIDI needs a secure context and ${here} is not one. `
                    + 'Open the rack on localhost, put it behind HTTPS, or '
                    + 'forward the port so it arrives as localhost: '
                    + '`ssh -L 8000:localhost:8000 <host>`. The LAN address '
                    + 'this server prints will not do on its own.',
            };
        }
        if (!navigator.requestMIDIAccess) {
            return {
                state: 'unsupported',
                detail: 'this browser exposes no Web MIDI, and the page is a '
                    + 'secure context - so it is the browser or its settings',
            };
        }
        return null;
    }

    // Whether the remedy is likely to be an ALSA one.
    //
    // Deliberately a hint rather than a branch in the logic: the browser cannot
    // see ALSA, so all this does is name the next command instead of leaving
    // "no inputs are attached" as a dead end.
    onLinux() {
        const platform = (typeof navigator !== 'undefined'
            && (navigator.userAgentData?.platform || navigator.platform)) || '';
        return /linux|arm/i.test(platform);
    }

    // granted | denied | prompt | unknown.
    //
    // Worth asking separately, because a browser that blocks MIDI by policy or
    // by a shield reports `denied` without ever prompting, and that is
    // indistinguishable from a dismissed prompt in the failure alone.
    async permissionState() {
        if (typeof navigator === 'undefined' || !navigator.permissions?.query) {
            return 'unknown';
        }
        try {
            const status = await navigator.permissions.query(
                { name: 'midi', sysex: false });
            return status.state;
        } catch {
            // Some browsers refuse to answer for `midi` at all, which is not
            // the same as refusing MIDI.
            return 'unknown';
        }
    }

    async connect() {
        const wrong = this.diagnose();
        if (wrong) {
            this.onStatus(wrong);
            return false;
        }
        try {
            this.access = await navigator.requestMIDIAccess({ sysex: false });
        } catch (error) {
            const permission = await this.permissionState();
            this.onStatus({
                state: 'refused',
                detail: permission === 'denied'
                    ? 'permission is denied for this site - check the browser\'s '
                      + 'MIDI site setting, and any shield or extension that '
                      + 'blocks device access'
                    : `${error.message} (permission: ${permission})`,
            });
            return false;
        }

        this.ports = [...this.access.inputs.values()];
        this.ports.forEach(port => {
            port.onmidimessage = (event) => this.receive(event.data, port.name);
        });
        this.access.onstatechange = () => this.rescan();

        this.onStatus({
            state: this.ports.length ? 'connected' : 'no-ports',
            detail: this.ports.length
                ? this.ports.map(p => p.name).join(', ')
                : 'Web MIDI is available and no inputs are attached'
                  + (this.onLinux()
                     ? ' - on Linux the browser reaches MIDI through ALSA, so '
                       + 'check `aconnect -l` first: a port ALSA cannot see is '
                       + 'invisible to every browser on the machine'
                     : ''),
        });
        return true;
    }

    rescan() {
        if (!this.access) return;
        const seen = new Set(this.ports.map(p => p.id));
        [...this.access.inputs.values()].forEach(port => {
            if (seen.has(port.id)) return;
            port.onmidimessage = (event) => this.receive(event.data, port.name);
            this.ports.push(port);
        });
    }

    // The one path every message takes, whichever side it came from. The
    // synthetic source below calls this too, so nothing is exercised only by
    // hardware nobody has plugged in.
    receive(bytes, from = 'synthetic') {
        let message;
        try {
            message = midiParse(bytes);
        } catch (error) {
            this.onStatus({ state: 'bad-message', detail: error.message });
            return [];
        }

        this.received += 1;
        this.lastMessage = { ...message, from };
        this.onMessage(message);
        const activity = midiRoute(this.bindings, message);
        this.matched += activity.length;
        activity.forEach(item => this.onActivity(item));
        return activity;
    }

    // A rack with no hardware attached still has to be demonstrable, and a
    // feature that can only be seen with a keyboard plugged in is a feature
    // nobody reviews.
    simulate({ channel = 1, note = 36, controller = null, value = 100 } = {}) {
        const bytes = controller === null
            ? [0x90 | (channel - 1), note, value]
            : [0xB0 | (channel - 1), controller, value];
        return this.receive(bytes, 'simulated');
    }
}

File: src/midi.py
"""MIDI in, and where it lands.

Carlos does not make sound, so a MIDI message here is not a note to play — it is
evidence that something in the rig did something, and the question this module
answers is *which device on screen was it*.

The seam is MIDI itself. It is a standard protocol with more independent
implementations than anyone can count, which is exactly what the QM seams record
asks of anything reaching a third party: a from-scratch implementation of the
protocol alone is enough to drive this, and nothing about Carlos needs to be
known on the other side.

Three pieces, all pure:

- `parse` turns bytes into a message. Real MIDI arrives as status + data bytes,
  and a build that only accepts a friendly JSON shape has not anticipated a MIDI
  signal, it has anticipated a description of one.
- `MidiBinding` says which messages belong to which device.
- `route` matches a message against the bindings and reports what lit up.

Nothing here opens a port. The browser's Web MIDI adapter and the HTTP endpoint
both call `route`, so the mapping behaves identically whichever side the message
came in on, and the tests need no hardware.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, model_validator

# Channel voice messages, by the high nibble of the status byte.
NOTE_OFF = 0x80
NOTE_ON = 0x90
POLY_AFTERTOUCH = 0xA0
CONTROL_CHANGE = 0xB0
PROGRAM_CHANGE = 0xC0
CHANNEL_AFTERTOUCH = 0xD0
PITCH_BEND = 0xE0

# System real-time, whole-byte statuses. RAD drives playback from these, so the
# same three bytes mean the same thing in both projects.
CLOCK = 0xF8
START = 0xFA
CONTINUE = 0xFB
STOP = 0xFC

# RAD claims CC 20-23 for its speed axes (bpm, aps, div, quantize). A binding
# here may still *watch* them, but Carlos must not give them a second meaning:
# the seams record permits extending a shared vocabulary and forbids
# repurposing it.
RAD_RESERVED_CC = (20, 21, 22, 23)

MessageType = Literal[
    "note_on", "note_off", "aftertouch", "cc",
    "program", "pitchbend", "clock", "start", "stop", "continue",
]

SourceType = Literal["channel", "note", "cc", "program", "pitchbend", "transport"]


class MidiError(ValueError):
    """Bytes that are not a message, or a binding that cannot be read."""


class MidiMessage(BaseModel):
    """One parsed message."""

    model_config = ConfigDict(extra="forbid")

    type: MessageType
    channel: int | None = Field(default=None, ge=1, le=16)
    note: int | None = Field(default=None, ge=0, le=127)
    controller: int | None = Field(default=None, ge=0, le=127)
    # Velocity for notes, value for CC, 0..16383 for pitch bend.
    value: int | None = Field(default=None, ge=0, le=16383)

    @property
    def intensity(self) -> float:
        """0..1, for something on screen to be brighter or dimmer.

        A note-off is zero however it was spelled, which matters because the
        two spellings below are both real and a display that honoured only one
        would leave a device lit forever.
        """
        if self.type in ("note_off", "stop"):
            return 0.0
        if self.type == "pitchbend":
            return (self.value or 0) / 16383
        if self.value is None:
            return 1.0
        return self.value / 127


def parse(data: "bytes | list[int]") -> MidiMessage:
    """Turn raw MIDI bytes into a message, or say why they are not one."""
    payload = list(data)
    if not payload:
        raise MidiError("no bytes")

    status = payload[0]
    if status < 0x80:
        raise MidiError(f"0x{status:02X} is a data byte, not a status byte")

    if status >= 0xF8:
        transport = {CLOCK: "clock", START: "start", STOP: "stop", CONTINUE: "continue"}
        if status not in transport:
            raise MidiError(f"unsupported system message 0x{status:02X}")
        return MidiMessage(type=transport[status])

    if status >= 0xF0:
        raise MidiError(f"system common 0x{status:02X} is not mapped")

    kind = status & 0xF0
    channel = (status & 0x0F) + 1  # wire is 0-based; humans count from 1
    data1 = payload[1] if len(payload) > 1 else None
    data2 = payload[2] if len(payload) > 2 else None

    def need(*values):
        if any(v is None for v in values):
            raise MidiError(f"0x{status:02X} is missing its data bytes")

    if kind == NOTE_ON:
        need(data1, data2)
        # A note-on with velocity 0 is a note-off. Devices really send this,
        # and treating it as a note-on leaves the display stuck lit.
        return MidiMessage(
            type="note_on" if data2 else "note_off",
            channel=channel, note=data1, value=data2,
        )

    if kind == NOTE_OFF:
        need(data1, data2)
        return MidiMessage(type="note_off", channel=channel, note=data1, value=data2)

    if kind == POLY_AFTERTOUCH:
        need(data1, data2)
        return MidiMessage(type="aftertouch", channel=channel, note=data1, value=data2)

    if kind == CONTROL_CHANGE:
        need(data1, data2)
        return MidiMessage(type="cc", channel=channel, controller=data1, value=data2)

    if kind == PROGRAM_CHANGE:
        need(data1)
        return MidiMessage(type="program", channel=channel, value=data1)

    if kind == CHANNEL_AFTERTOUCH:
        need(data1)
        return MidiMessage(type="aftertouch", channel=channel, value=data1)

    if kind == PITCH_BEND:
        need(data1, data2)
        # 14-bit, LSB first.
        return MidiMessage(type="pitchbend", channel=channel, value=(data2 << 7) | data1)

    raise MidiError(f"unhandled status 0x{status:02X}")


def encode(message: "MidiMessage | dict") -> list[int]:
    """Turn a message into the bytes a device will receive, or refuse it.

    The inverse of `parse`, and the only way anything leaves this application.
    Every send goes through here, so a malformed message is refused with the
    reason rather than handed to a device that will report it as garbage - which
    is the failure this exists for: a device saying "malformed" tells you
    something was wrong and never which byte.

    The refusal is deliberately stricter than the wire. MIDI itself cannot
    express a note of 200 - the eighth bit marks a status byte - so a value out
    of range does not produce a bad note, it produces a byte the device reads as
    a new message and then starves for data. That is why a single bad value can
    desynchronise a stream rather than sound wrong.

    `parse(encode(m)) == m` for every message this can build, which is the
    property the tests assert rather than a table of expected bytes.
    """
    if not isinstance(message, MidiMessage):
        try:
            message = MidiMessage.model_validate(message)
        except ValidationError as error:
            # The same shaping `load_binding` does below: pydantic's own string
            # carries the model name, an echo of the input and a versioned docs
            # URL, none of which help somebody holding a message that will not
            # send.
            problems = "; ".join(
                f"{'.'.join(str(part) for part in item['loc'])}: "
                f"{item['msg'].removeprefix('Value error, ')}"
                for item in error.errors()
            )
            raise MidiError(problems) from None

    kind = message.type

    transport = {"clock": CLOCK, "start": START, "stop": STOP, "continue": CONTINUE}
    if kind in transport:
        return [transport[kind]]

    if message.channel is None:
        raise MidiError(f"a {kind} message needs a channel")
    status_channel = message.channel - 1

    def seven(name: str, value: "int | None", default: "int | None" = None) -> int:
        if value is None:
            if default is None:
                raise MidiError(f"a {kind} message needs {name}")
            value = default
        if not 0 <= value <= 127:
            raise MidiError(
                f"{name} is {value}; MIDI carries 0-127 in a data byte, and "
                f"anything above it sets the bit that marks a status byte"
            )
        return value

    if kind == "note_on":
        return [NOTE_ON | status_channel,
                seven("a note", message.note),
                seven("a velocity", message.value, 64)]

    if kind == "note_off":
        # Sent as a real note-off rather than note-on-with-zero. Both are legal
        # and devices differ on which they prefer; the explicit one is never
        # mistaken for a note that failed to sound.
        return [NOTE_OFF | status_channel,
                seven("a note", message.note),
                seven("a velocity", message.value, 0)]

    if kind == "cc":
        return [CONTROL_CHANGE | status_channel,
                seven("a controller", message.controller),
                seven("a value", message.value, 0)]

    if kind == "program":
        return [PROGRAM_CHANGE | status_channel, seven("a program", message.value)]

    if kind == "aftertouch":
        # Polyphonic when it names a note, channel-wide when it does not.
        if message.note is not None:
            return [POLY_AFTERTOUCH | status_channel,
                    seven("a note", message.note),
                    seven("a pressure", message.value, 0)]
        return [CHANNEL_AFTERTOUCH | status_channel, seven("a pressure", message.value)]

    if kind == "pitchbend":
        value = message.value if message.value is not None else 8192
        if not 0 <= value <= 16383:
            raise MidiError(
                f"a pitch bend is {value}; it is 14 bits, so 0-16383, "
                f"sent as two 7-bit bytes"
            )
        return [PITCH_BEND | status_channel, value & 0x7F, (value >> 7) & 0x7F]

    raise MidiError(f"nothing knows how to send a {kind} message")


class MidiSource(BaseModel):
    """Which messages a binding is interested in.

    A field left out is a wildcard, so `{type: channel, channel: 10}` is "any
    message on channel 10" and `{type: note, note: 36}` is "kick, on any
    channel". Omitting the channel is deliberate rather than lazy: a rig where
    one machine is moved to another channel should not need its bindings
    rewritten.
    """

    model_config = ConfigDict(extra="forbid")

    type: SourceType
    channel: int | None = Field(default=None, ge=1, le=16)
    note: int | None = Field(default=None, ge=0, le=127)
    controller: int | None = Field(default=None, ge=0, le=127)

    @model_validator(mode="after")
    def _selector_matches_type(self) -> "MidiSource":
        if self.type == "note" and self.controller is not None:
            raise ValueError("a note source cannot select a controller")
        if self.type == "cc" and self.note is not None:
            raise ValueError("a cc source cannot select a note")
        return self


class MidiBinding(BaseModel):
    """One rule: messages like *this* belong to *that* device.

    `parameter` and `jack` are both optional and are about what the device does
    when it is hit — turn a knob, light a socket. A binding with neither still
    lights the device, which is the common case and the reason neither is
    required.
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    source: MidiSource
    module: str = Field(min_length=1)
    parameter: str | None = None
    jack: str | None = None
    label: str = ""


class Activity(BaseModel):
    """What one message did to the picture."""

    model_config = ConfigDict(extra="forbid")

    binding: str
    module: str
    parameter: str | None = None
    jack: str | None = None
    intensity: float
    message: MidiMessage


def matches(source: MidiSource, message: MidiMessage) -> bool:
    """Whether a message is one this source asked for."""
    if source.channel is not None and message.channel != source.channel:
        return False

    if source.type == "channel":
        return message.channel is not None

    if source.type == "note":
        if message.type not in ("note_on", "note_off", "aftertouch"):
            return False
        return source.note is None or message.note == source.note

    if source.type == "cc":
        if message.type != "cc":
            return False
        return source.controller is None or message.controller == source.controller

    if source.type == "program":
        return message.type == "program"

    if source.type == "pitchbend":
        return message.type == "pitchbend"

    if source.type == "transport":
        return message.type in ("clock", "start", "stop", "continue")

    return False


def route(bindings: "list[MidiBinding]", message: MidiMessage) -> "list[Activity]":
    """Every device this message lights, in binding order.

    All matching bindings fire, rather than the most specific one winning. One
    message genuinely can concern two devices — a clock lights everything
    following it — and picking a single winner would hide that. A rig where two
    bindings fight over one parameter is a mapping to fix, not a tie for this
    function to break silently.
    """
    return [
        Activity(
            binding=binding.id,
            module=binding.module,
            parameter=binding.parameter,
            jack=binding.jack,
            intensity=message.intensity,
            message=message,
        )
        for binding in bindings
        if matches(binding.source, message)
    ]


def load_binding(document: object) -> MidiBinding:
    if not isinstance(document, dict):
        raise MidiError(f"a binding is an object, not {type(document).__name__}")
    try:
        return MidiBinding.model_validate(document)
    except ValidationError as exc:
        problems = "; ".join(
            f"{'.'.join(str(p) for p in d['loc'])}: "
            f"{d['msg'].removeprefix('Value error, ')}"
            for d in exc.errors()
        )
        raise MidiError(problems) from exc
